test/unit/math/rev/mat/functor/algebra_solver_test.cpp

```diff
@@ -312,3 +339,205 @@ TEST(MathMatrix, system_functor_constructor) {
+        = non_linear_eq_test(non_linear_eq_functor(), y, is_newton);
+
+    EXPECT_FLOAT_EQ(-y(0).val(), theta(0).val());
+    EXPECT_FLOAT_EQ(-y(1).val(), theta(1).val());
+    EXPECT_FLOAT_EQ(y(2).val(), theta(2).val());
+
+    AVEC y_vec = createAVEC(y(0), y(1), y(2));
+    VEC g;
+    theta(k).grad(y_vec, g);
+
+    for (int i = 0; i < n_y; i++)
+      EXPECT_NEAR(J(k, i), g[i], err);
+  }
+}
+
+TEST_F(algebra_solver_non_linear_eq_test, newton_dbl) {
+  bool is_newton = true;
+  Eigen::VectorXd theta
+      = non_linear_eq_test(non_linear_eq_functor(), y_dbl, is_newton);
+  EXPECT_FLOAT_EQ(-y_dbl(0), theta(0));
+  EXPECT_FLOAT_EQ(-y_dbl(1), theta(1));
+  EXPECT_FLOAT_EQ(y_dbl(2), theta(2));
+}
+
+TEST_F(error_message_test, newton) {
+  using stan::math::var;
+  bool is_newton = true;
+  Eigen::Matrix<var, Eigen::Dynamic, 1> y = y_2;
+  error_conditions_test(non_linear_eq_functor(), y, is_newton);
+}
+
+TEST_F(error_message_test, newton_dbl) {
+  bool is_newton = true;
+  error_conditions_test(non_linear_eq_functor(), y_3, is_newton);
+}
+
+TEST_F(max_steps_test, newton) {
+  bool is_newton = true;
+  max_num_steps_test(y_var, is_newton);
+}
+
+TEST_F(max_steps_test, newton_dbl) {
+  bool is_newton = true;
+  max_num_steps_test(y, is_newton);
+}
+
+TEST(MathMatrix, unsolvable_flag_newton) {
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> y(2);
+  y << 1, 1;
+
+  unsolvable_flag_test(y);
+}
+
+TEST(MathMatrix, unsolvable_flag_newton_dbl) {
+  Eigen::VectorXd y(2);
+  y << 1, 1;
+
+  unsolvable_flag_test(y);
+}
+
+TEST_F(degenerate_eq_test, newton) {
+  using stan::math::algebra_solver_powell;
+  using stan::math::sum;
+  using stan::math::var;
+
+  // This first initial guess produces the
+  // solution x = {8, 8}
+  bool is_newton = true;
+  for (int k = 0; k < n_x; k++) {
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y = y_dbl;
+    Eigen::Matrix<var, Eigen::Dynamic, 1> theta
+        = degenerate_test(y, x_guess_1, is_newton);
+    EXPECT_FLOAT_EQ(8, theta(0).val());
+    EXPECT_FLOAT_EQ(8, theta(1).val());
+
+    AVEC y_vec = createAVEC(y(0), y(1));
+    VEC g;
+    theta(k).grad(y_vec, g);
+
+    for (int l = 0; l < n_y; l++)
+      EXPECT_NEAR(J1(k, l), g[l], tolerance);
+  }
+
+  // This next initial guess produces the
+  // solution x = {5, 5}
+  for (int k = 0; k < 1; k++) {
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y = y_dbl;
+    Eigen::Matrix<var, Eigen::Dynamic, 1> theta
+        = degenerate_test(y, x_guess_2, is_newton);
+    EXPECT_FLOAT_EQ(5, theta(0).val());
+    EXPECT_FLOAT_EQ(5, theta(0).val());
+
+    AVEC y_vec = createAVEC(y(0), y(1));
+    VEC g;
+    theta(k).grad(y_vec, g);
+
+    for (int l = 0; l < n_y; l++)
+      EXPECT_NEAR(J2(k, l), g[l], tolerance);
+  }
+}
+
+TEST_F(degenerate_eq_test, newton_dbl) {
+  bool is_newton = true;
+
+  // This first initial guess produces the
+  // solution x = {8, 8}
+  Eigen::VectorXd theta = degenerate_test(y_dbl, x_guess_1, is_newton);
```

Comment:
There are unnecessarily amount complexity for a single test. This three layers of
`degenerate_test` -> `general_algebra_solver` -> `algebra_solver_newton` can simply be
```c++
algebra_solver_newton(degenerate_eq_functor(), x_guess_1, y_dbl, 
                      std::vector<double>(), std::vector<int>());
```
The three-func-call version makes unit test less readable and much error-prone. For example, `degenerate_test` has `x` and `y` in opposite order with `algebra_solver()`, increasing chance of bugs. Please just use `algebra_solver_powell` and `algebra_solver_newton` unless it's absolutely necessary for additional layers.

Reply:
Agreed. Easy fix.

Usually, the `*_test()` functions do a few things, hence their utility, but `degenerate_test()` only adds an extra layer. I think this scheme made more sense in some earlier design of the unit test.

Reply:
Fixed.